src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.cs

```diff
@@ -952,90 +969,98 @@ private struct CurrentState
+            public static int FixedLength(SymbolicRegexBuilder<TSet> builder, ref CurrentState state, uint nextCharKind) => state.DfaState(builder)!.FixedLength(nextCharKind);
 
             /// <summary>Take the transition to the next DFA state.</summary>
             [MethodImpl(MethodImplOptions.AggressiveInlining)]
             public static bool TakeTransition(SymbolicRegexBuilder<TSet> builder, ref CurrentState state, int mintermId)
             {
-                Debug.Assert(state.DfaState is not null, $"Expected non-null {nameof(state.DfaState)}.");
+                Debug.Assert(state.DfaStateId > 0, $"Expected non-zero {nameof(state.DfaStateId)}.");
                 Debug.Assert(state.NfaState is null, $"Expected null {nameof(state.NfaState)}.");
                 Debug.Assert(builder._delta is not null);
 
-                // Get the current state.
-                DfaMatchingState<TSet> dfaMatchingState = state.DfaState!;
-
                 // Use the mintermId for the character being read to look up which state to transition to.
                 // If that state has already been materialized, move to it, and we're done. If that state
                 // hasn't been materialized, try to create it; if we can, move to it, and we're done.
-                int dfaOffset = (dfaMatchingState.Id << builder._mintermsLog) | mintermId;
-                DfaMatchingState<TSet>? nextState = builder._delta[dfaOffset];
-                if (nextState is not null || builder.TryCreateNewTransition(dfaMatchingState, mintermId, dfaOffset, checkThreshold: true, out nextState))
+                int dfaOffset = (state.DfaStateId << builder._mintermsLog) | mintermId;
+                int nextStateId = builder._delta[dfaOffset];
+                if (nextStateId > 0 || builder.TryCreateNewTransition(state.DfaStateId, mintermId, dfaOffset, checkThreshold: true, out nextStateId))
                 {
                     // There was an existing state for this transition or we were able to create one.  Move to it and
                     // return that we're still operating as a DFA and can keep going.
-                    state.DfaState = nextState;
+                    state.DfaStateId = nextStateId;
                     return true;
                 }
 
                 return false;
             }
+
+            /// <summary>
+            /// Gets context independent state information:
+            /// - whether this is an initial state
+            /// - whether this is a dead-end state, meaning there are no transitions possible out of the state
+            /// - whether this state is unconditionally nullable
+            /// - whether this state may be contextually nullable
+            /// </summary>
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static (bool IsInitial, bool IsDeadend, bool IsNullable, bool CanBeNullable) GetStateInfo(SymbolicRegexBuilder<TSet> builder, ref CurrentState state)
```

Comment:
I did have the thought too. That would save memory, so worth a try.

Reply:
Perhaps it could just be a ulong extension of `SymbolicRegexInfo._info` (that is a `uint` wrapped inside the struct but it uses all the 8 bits already, essentially I think 2 more bit are needed for `IsInitial`, and `IsDeadend` as state info, other bits are already in the node info itself)

Reply:
Oh, there are actually 8 more bits available in a `uint` itslef (mixed it up with `ushort` for a sec)